packages/router/router.ts

```diff
@@ -836,14 +870,64 @@ export function createRouter(init: RouterInit): Router {
       }
     );
 
-    // Kick off initial data load if needed.  Use Pop to avoid modifying history
-    if (!state.initialized) {
+    if (state.initialized) {
+      return router;
+    }
+
+    let lazyMatches = state.matches.filter((m) => m.route.lazy);
+
+    if (lazyMatches.length === 0) {
+      // Kick off initial data load if needed.  Use Pop to avoid modifying history
       startNavigation(HistoryAction.Pop, state.location);
+      return router;
     }
 
+    // Load lazy modules, then kick off initial data load if needed
+    let lazyPromises = lazyMatches.map((m) =>
+      loadLazyRouteModule(m.route, hasErrorBoundary, manifest)
+    );
+    Promise.all(lazyPromises).then(() => {
+      let initialized =
+        !state.matches.some((m) => m.route.loader) ||
+        init.hydrationData != null;
+      if (initialized) {
+        // We already have required loaderData so we can just set initialized
+        updateState({ initialized: true });
+      } else {
+        // We still need to kick off initial data loads
+        startNavigation(HistoryAction.Pop, state.location);
+      }
+    });
+
     return router;
   }
 
+  // Returns a promise that resolves when the router has been initialized
+  // including any lazy-loaded route properties.  This is useful on the client
+  // after server-side rendering to ensure that the routes are ready to render
+  // since all elements and error boundaries have been resolved.
+  //
+  // Implemented as a Fluent API for ease of: let router = await
+  //   createRouter(init).initialize().ready();
+  //
+  // TODO: Rename this and/or initialize()?  If we make initialize() async then
+  // the public router creation functions will become async too which is a
+  // breaking change.
+  function ready(): Promise<Router> {
+    return new Promise((resolve) => {
```

Comment:
Flagging for review — this promise is never rejected. What's the expected flow when hydrating on the client but a lazy route fails to load? Right now we get a server/client mismatch because the server successfully resolved the lazy route but the client renders an error message.

Reply:
Tabled `ready()` for now since it can be handled in userland